diesel/src/query_dsl/combine_dsl.rs

```diff
@@ -0,0 +1,246 @@
+use crate::query_builder::combination_clause::{
+    All, Combination, Distinct, Except, Intersect, Union,
+};
+use crate::query_builder::{AsQuery, BoxedSelectStatement, Query, SelectStatement};
+use crate::Table;
+
+/// Extension trait to combine queries using a combinator like `UNION`, `INTERSECT` or `EXPECT`
+/// with or without `ALL` rule for duplicates
+pub trait CombineDsl {
+    /// What kind of query does this type represent?
+    type Query: Query;
+
+    /// Combine two queries using a SQL `UNION`
+    ///
+    /// # Examples
+    /// ```rust
+    /// # #[macro_use] extern crate diesel;
+    /// # include!("../doctest_setup.rs");
+    /// # use schema::{users, animals};
+    ///
+    /// # fn main() {
+    /// #     use self::users::dsl::{users, name as user_name};
+    /// #     use self::animals::dsl::{animals, name as animal_name};
+    /// #     let connection = establish_connection();
+    /// let data = users.select(user_name.nullable())
+    ///     .union(animals.select(animal_name))
+    ///     .load(&connection);
+    ///
+    /// let expected_data = vec![
+    ///     None,
+    ///     Some(String::from("Jack")),
+    ///     Some(String::from("Tess")),
+    ///     Some(String::from("Sean")),
+    /// ];
+    /// assert_eq!(Ok(expected_data), data);
+    /// # }
+    /// ```
+    fn union<Rhs>(self, rhs: Rhs) -> Combination<Union, Distinct, Self::Query, Rhs::Query>
```

Comment:
`diesel::dsl` should reexport helper types for each of those methods, so that we can write the return type here as `diesel::dsl::Union<Self, Rhs>` or `diesel::dsl::UnionAll<Self, Rhs>` (This allows potential users to refer easier to those types)